Make LOCK, LOCK2, TRY_LOCK work with CWaitableCriticalSection

ryanofsky opened the pull request:
Make LOCK macros work with non-recursive mutexes, and use wherever possible for better deadlock detection.

Also add unit test for DEBUG_LOCKORDER code.

ryanofsky committed:
Add unit test for DEBUG_LOCKORDER code

ryanofsky committed:
Make LOCK, LOCK2, TRY_LOCK work with CWaitableCriticalSection

They should also work with any other mutex type which std::unique_lock
supports.

There is no change in behavior for current code that calls these macros with
CCriticalSection mutexes.

ryanofsky committed:
Use LOCK macros for non-recursive locks

Instead of std::unique_lock.

TheBlueMatt commented on src/sync.cpp:
```diff
@@ -95,7 +95,7 @@ static void potential_deadlock_detected(const std::pair<void*, void*>& mismatch,
         LogPrintf(" %s\n", i.second.ToString());
     }
-    assert(false);
+    throw std::logic_error("potential deadlock detected");
```
I believe we would sometimes incorrectly catch this - we still have some generic catch blocks lying around in places :(. I wouldn't mind having a policy of only catching in tests, but until then, I think this should, sadly, remain an assert (unless you want to go to the hassle of adding some #define that is only set in test_bitcoin and compiling this unit differently for it :/).

ryanofsky replied:
What exactly is the problem? DEBUG_LOCKORDER isn't enabled unless you are doing a special build anyway.

TheBlueMatt replied:
I presume it wouldn't actually fail in some cases where a lockorder inversion is hit as the exception here would be caught. Precisely because its (essentially) only enabled in travis we should just keep an assert.

ryanofsky replied:
> I presume it wouldn't actually fail in some cases where a lockorder inversion is hit as the exception here would be caught. Precisely because its (essentially) only enabled in travis we should just keep an assert.

Makes sense, added back an optional abort (on by default), so it will still be guaranteed to fail travis, even in the presence of code that would catch std::logic_error exceptions.

TheBlueMatt reviewed (commented):
Dont you need a similar DeleteLock() call as CCriticalSection?

TheBlueMatt commented:
Fails travis due to new lock checking in clang.

ryanofsky commented:
> Dont you need a similar DeleteLock() call as CCriticalSection?

Good catch, added this.

TheBlueMatt commented on src/sync.h:
```diff
@@ -93,73 +95,72 @@ void static inline DeleteLock(void* cs) {}
 {
 public:
-    ~CCriticalSection() {
+    ~LockOrderMixin() {
```
Maybe just move this into AnnotatedMixin instead of making everything two classes on top of each other?

TheBlueMatt commented on src/sync.h:
```diff
@@ -168,22 +169,27 @@ class SCOPED_LOCKABLE CCriticalBlock
-#define TRY_LOCK(cs, name) CCriticalBlock name(cs, #cs, __FILE__, __LINE__, true)
+#define LOCK(cs) DebugLock<decltype(cs)> PASTE2(criticalblock, __COUNTER__)(cs, #cs, __FILE__, __LINE__)
+#define LOCK2(cs1, cs2)                                               \
+    DebugLock<decltype(cs1)> criticalblock1(cs1, #cs1, __FILE__, __LINE__); \
```
Isnt it considered bad practice to turn a #define into two statements (asking, I dont know, I've just always seen people bend over to avoid it).

TheBlueMatt replied:
Fair enough, I just missed why you changed it, but see it now.

TheBlueMatt reviewed (commented):
Awesome, thanks for tackling this.

ryanofsky committed:
MOVEONLY Move AnnotatedMixin declaration

Move AnnotatedMixin closer to where it's used, and after the DEBUG_LOCKORDER
function declarations so it can call them.

ryanofsky reviewed (commented):
Added 2 commits 6e3d96e0aae2040dc0d92adf02c901522bf28289 -> 396911e0e15129c218634c868acb747e423e88d9 ([pr/dead.5](https://github.com/ryanofsky/bitcoin/commits/pr/dead.5) -> [pr/dead.6](https://github.com/ryanofsky/bitcoin/commits/pr/dead.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/dead.5...pr/dead.6))
Squashed 396911e0e15129c218634c868acb747e423e88d9 -> f190dd6e5ebf07a75559855d04659502f6e63123 ([pr/dead.6](https://github.com/ryanofsky/bitcoin/commits/pr/dead.6) -> [pr/dead.7](https://github.com/ryanofsky/bitcoin/commits/pr/dead.7))

ryanofsky commented:
Rebased f190dd6e5ebf07a75559855d04659502f6e63123 -> 674dcfea1d0a5314e39fc9731cb863995ff34ea1 ([pr/dead.7](https://github.com/ryanofsky/bitcoin/commits/pr/dead.7) -> [pr/dead.8](https://github.com/ryanofsky/bitcoin/commits/pr/dead.8)) due to new changes conflicting with AssertLockNotHeld declaration from #10286.

TheBlueMatt commented:
utACK 674dcfea1d0a5314e39fc9731cb863995ff34ea1

ryanofsky commented:
Rebased 674dcfea1d0a5314e39fc9731cb863995ff34ea1 -> b6c88a148db170367abec036b6c4f2e13104f9aa ([pr/dead.8](https://github.com/ryanofsky/bitcoin/commits/pr/dead.8) -> [pr/dead.9](https://github.com/ryanofsky/bitcoin/commits/pr/dead.9)) due to with trivial conflicts with #12366 and #12367.

laanwj commented on src/httpserver.cpp:
```diff
@@ -115,7 +115,7 @@ class WorkQueue
     /** Interrupt and exit loops */
     void Interrupt()
     {
-        std::unique_lock<std::mutex> lock(cs);
```
I don't like replacing the standard C++11 lock usage here, and in other places, with the macros. The original code seems easier to understand at least.

ryanofsky replied:
> I don't like replacing the standard C++11 lock usage here, and in other places, with the macros

It doesn't matter to me which style is used, but the macros provide lock order checking which @TheBlueMatt seemed to think was important. It would be helpful if there could be a guideline for when to use LOCK macros vs standard locks. Is the guideline that you'd recommend just to avoid macros when locks are used with condition variables, or is it broader or narrower than that?

TheBlueMatt replied:
We're in a not-great state right now where we have automated lock-checking code, but a bunch of places dont bother to use it. In practice most of those places are really limited use and "obviously-correct" (tm), but that only makes it much easier to break in the future, and making exceptions like that are kinda bad. I would strongly prefer we have a consistent syntax across the codebase that uses our lock-checking stuff vs moving back to an ad-hoc world.

ryanofsky replied:
> I would strongly prefer we have a consistent syntax across the codebase that uses our lock-checking stuff vs moving back to an ad-hoc world.

This PR switches code to consistently use LOCK syntax, which Wladimir objects to in some cases (not clear which ones). 

@laanwj @TheBlueMatt, it'd be good to resolve the apparent disagreement about when to use `lock_guard` and when to use `LOCK`, if possible. But if there can't be any agreement, I could revert the changes in the PR to keep preexisting styles.

laanwj replied:
I agree the lock checking is important, but I dislike the non-standard macro syntax as it hides what is happening. So I wondered if there is an official way to 'decorate' the C++11 locking primitives, as I doubt we're the only project to do checking like this. If not, this change is fine with me...

TheBlueMatt commented:
Yes, I tend to agree with that point. I think just explicitly using our wrapper classes instead of the LOCK macros would suffice there.

On March 2, 2018 6:08:09 PM UTC, "Wladimir J. van der Laan" <notifications@github.com> wrote:
>laanwj commented on this pull request.
>
>
>
>> @@ -115,7 +115,7 @@ class WorkQueue
>     /** Interrupt and exit loops */
>     void Interrupt()
>     {
>-        std::unique_lock<std::mutex> lock(cs);
>
>I agree the lock checking is important, but I dislike the non-standard
>macro syntax as it hides what is happening. So I wondered if there is
>an official way to 'decorate' the C++11 locking primitives, as I doubt
>we're the only project to do checking like this. If not, this change is
>fine with me...
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/bitcoin/bitcoin/pull/11640#discussion_r171920098

ryanofsky commented:
> Yes, I tend to agree with that point. I think just explicitly using our wrapper classes instead of the LOCK macros would suffice there.

Matt, maybe I'm misunderstanding, but are you suggesting that I write something like:

```c++
DebugLock<CWaitableCriticalSection> lock(cs, "cs", __FILE__, __LINE__);
```

to replace current master code:

```c++
std::unique_lock<std::mutex> lock(cs);
```

instead of current PR code:

```c++
LOCK(cs);
```

If this is the case, I guess I'd prefer to keep the current PR code, since it just uses one style (`LOCK`) uniformly and it seems Wladimir no longer objects (https://github.com/bitcoin/bitcoin/pull/11640#discussion_r171920098). In the future, if we want to avoid using `LOCK` in some cases without replacing it everywhere, having a documented developer guideline would probably be helpful.

laanwj commented:
Yes, I agree, my comment was somehow out of scope. Let's leave changing the style to something in the future. Or maybe just documenting how the various OS/C++ synchronization primitives map to "bitcoin core style".

ryanofsky commented:
Rebased b6c88a148db170367abec036b6c4f2e13104f9aa -> 41dea6a481781e69a1e404a7e3d806f59b47032e ([pr/dead.9](https://github.com/ryanofsky/bitcoin/commits/pr/dead.9) -> [pr/dead.10](https://github.com/ryanofsky/bitcoin/commits/pr/dead.10)) due to conflict with #12926

MarcoFalke commented on src/test/sync_tests.cpp:
```diff
@@ -0,0 +1,52 @@
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "sync.h"
+#include "test/test_bitcoin.h"
```
Please use ` #include <sync.h> ` instead.

laanwj commented:
utACK 9c4dc597ddc66acfd58a945a5ab11f833731abba
